pkg/kms/localkms/localkms_writer.go

```diff
@@ -17,21 +18,82 @@ import (
 const maxKeyIDLen = 20
 
 // newWriter creates a new instance of local storage key storeWriter in the given store and for masterKeyURI
-func newWriter(kmsStore storage.Store) *storeWriter {
+func newWriter(kmsStore storage.Store, opts ...PrivateKeyOpts) *storeWriter {
+	pOpts := &privateKeyOpts{}
+
+	for _, opt := range opts {
+		opt(pOpts)
+	}
+
+	reqID := ""
+	if opts != nil && pOpts.ksID != "" {
+		reqID = pOpts.ksID
+	}
+
 	return &storeWriter{
-		storage: kmsStore,
+		storage:           kmsStore,
+		requestedKeysetID: reqID,
 	}
 }
 
 // storeWriter struct to store a keyset in a local store
 type storeWriter struct {
 	storage storage.Store
+	//
+	requestedKeysetID string
 	// KeysetID is set when Write() is called
 	KeysetID string
 }
 
 // Write a marshaled keyset p in localstore with masterKeyURI prefix + randomly generated KeysetID
 func (l *storeWriter) Write(p []byte) (int, error) {
+	var err error
+
+	ksID := ""
+
+	if l.requestedKeysetID != "" {
+		ksID, err = l.verifyRequestedID()
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		ksID, err = l.newKeysetID()
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	err = l.storage.Put(ksID, p)
+	if err != nil {
+		return 0, err
+	}
+
+	l.KeysetID = ksID
+
+	return len(p), nil
+}
+
+func (l *storeWriter) verifyRequestedID() (string, error) {
+	if len(l.requestedKeysetID) > maxKeyIDLen {
+		return "", fmt.Errorf("requested ID '%s' is longer than max allowed length of %d", l.requestedKeysetID,
+			maxKeyIDLen)
+	}
+
+	_, err := l.storage.Get(l.requestedKeysetID)
+	if err == nil {
+		return "", fmt.Errorf("requested ID '%s' already exists, cannot write keyset", l.requestedKeysetID)
+	}
+
+	if err != nil {
+		if err != storage.ErrDataNotFound {
+			return "", fmt.Errorf("got error while verifying requested ID: %w", err)
+		}
+	}
+
+	return l.requestedKeysetID, nil
+}
```

Comment:
better to avoid nested conditions
```
func (l *storeWriter) verifyRequestedID() (string, error) {
    if len(l.requestedKeysetID) > maxKeyIDLen {
        return "", fmt.Errorf("requested ID '%s' is longer than max allowed length of %d", l.requestedKeysetID,
            maxKeyIDLen)
    }

    _, err := l.storage.Get(l.requestedKeysetID)
    if errors.Is(err, storage.ErrDataNotFound) {
        return l.requestedKeysetID, nil
    }

    if err != nil {
        return "", fmt.Errorf("got error while verifying requested ID: %w", err)
    }

    return "", fmt.Errorf("requested ID '%s' already exists, cannot write keyset", l.requestedKeysetID)
}
```

Reply:
updated